server/broadcast.go

```diff
@@ -488,35 +492,16 @@ func shouldStopSession(err error) bool {
 
 func verify(verifier *verification.SegmentVerifier, cxn *rtmpConnection,
 	sess *BroadcastSession, source *stream.HLSSegment,
-	res *net.TranscodeData, URIs []string) error {
+	res *net.TranscodeData, URIs []string, segData [][]byte) error {
 
-	// Cache segment contents if necessary.
+	// Cache segment contents
 	// If we need to retry transcoding because verification fails,
 	// the the segments' OS location will be overwritten.
 	// Cache the segments so we can restore them in OS if necessary.
-	renditionData := make([][]byte, len(URIs))
-	for i, fname := range URIs {
-		if sess.BroadcasterOS.IsExternal() && drivers.IsOwnExternal(fname) {
-			// If broadcaster is using external storage and segments are there
-			// Then have the verifier use that external storage.
-			continue
-		}
-
-		// Sanity check for absolute URIs to ensure we're actually local.
-		uri, err := url.ParseRequestURI(fname)
-		if err != nil {
-			return err // Implies this is retryable?
-		}
-		memOS, ok := sess.BroadcasterOS.(*drivers.MemorySession)
-		if !uri.IsAbs() && ok {
-			data := memOS.GetData(fname)
-			if data == nil {
-				return errors.New("Missing Local Data")
-			}
-			renditionData[i] = data
-		} else {
-			return errors.New("Expected local storage but did not have it")
-		}
+	renditionData := make([][]byte, len(segData))
+	for i := range segData {
+		renditionData[i] = make([]byte, len(segData[i]))
+		copy(renditionData[i], segData[i])
```

Comment:
Can't we just do `renditionData[i] = segData[i]` ?

Reply:
I originally copied the data in `segData[i]` because changes to `segData[i]` would result in changes to `params.Renditions[i]` since they would point to the same slice header. But, `segData` is not used anywhere in `server/broadcast.go` besides in `verify()` so it should be safe to set `params.Renditions` to `segData`. In 75f9d11, I updated a few tests to allocate a new slice for `segData` whenever passing it as an argument to `verify()` (otherwise, some tests would fail because mutations to the same slice for `segData` would mutate the values for `params.Renditions`).

Reply:
Ahh I see, I would be okay with not mutating as a good practice here as well. I'll leave it up to you.

Reply:
Going to keep this change as avoiding the copy of segment data (which can be largeish) is nice.

Reply:
Ah yes didn't think about the size, good point.